zookeeper-client/zookeeper-client-c/src/zookeeper.c

```diff
@@ -2812,6 +2844,11 @@ static void finalize_session_establishment(zhandle_t *zh) {
     LOG_DEBUG(LOGCALLBACK(zh), "Calling a watcher for a ZOO_SESSION_EVENT and the state=ZOO_CONNECTED_STATE");
     zh->input_buffer = 0; // just in case the watcher calls zookeeper_process() again
     PROCESS_SESSION_EVENT(zh, zh->state);
+
+    if (zh->sasl_client) {
+        /* some packets might have been delayed during SASL negotiaton. */
+        adaptor_send_queue(zh, 0);
+    }
```

Comment:
This did not show up during testing, but I have added it in my last push: as far as I can tell, nothing guarantees delayed packets will be flushed before the next ping without it.

(Cc: @eolivelli)

Reply:
makes sense